operator/v1/types.go

```diff
@@ -53,6 +53,18 @@ type ClusterManagerSpec struct {
 	// +optional
 	// +kubebuilder:default={mode: Default}
 	DeployOption ClusterManagerDeployOption `json:"deployOption,omitempty"`
+
+	// RegistrationConfiguration contains the configuration of registration
+	// +optional
+	RegistrationConfiguration *RegistrationConfiguration `json:"registrationConfiguration,omitempty"`
+}
+
+type RegistrationConfiguration struct {
+	// FeatureGates represents the list of feature gates for registration
+	// If it is set empty, default feature gates will be used.
+	// If it is set not empty, registration controller will use all items of it as args. And the default feature gates will be overrode if set.
+	// +optional
+	FeatureGates map[string]bool `json:"featureGates,omitempty"`
```

Comment:
What is the behavior if the featuregate keys specified (the strings) are invalid or not recognized?

This will effectively turn your feature gates into an API. Can you walk me through how an example like this from kube would work on this struct.
1. release 5, featuregate/Foo does not exist.  Kube does not allow specifying featuregate/Foo=false
2. release 6, featuregate/Foo exists and is false by default.  It is now possible to set featuregate/Foo=[false|true]
3. release 7, featuregate/Foo exists and is true by default.  If a cluster-admin upgrading from 6 to seven wants to continue having featuregate/Foo=false, he can set featuregate/Foo=false before upgrading.  Let's say the cluster-admin wants featuregate/Foo=false.
4. release 8, featuregate/Foo exists, is set to true, and can no longer be set to false. Setting featuregate/Foo=false causes the kube-apiserver to exit.  In an HA upgrade, this allows determination during testing in test or stage clusters and rollback. This is preferable to having the cluster-admin intent not honored.

Using this API, how would similar feedback be provided back to the person setting this value.

Reply:
a working example i can think of is:

1. in release5 where featuregate/`Foo` doesnt exist, the registration-operator will skip configuring `Foo` for the components and set a degrade condition in the status of `ClusterManager`/`Klusterlet`.
2. in release6 where featuregate/`Foo` is under alpha phase (exists but defaulted to disabled), setting `Foo=[true|false]` in the operator api will override the defaults.
3. in release7 where featuregate/`Foo` is beta or GA (exists and defaulted to enabled), setting `Foo=[true|false]` in the operator api will override the defaults. e.g. explicitly disabling a GA featuregate means that the admin wants `Foo`=false.
4. in release8 where featuregate/`Foo` is locked to true, the configured setting in the operator api will be skipped w/ a degrade condition reporting the conflict.

Reply:
also, i have another concern about configuring the `--feature-gate` args per-component. say the work has a feature gate `Foo` while the registration doesnt have it, what's the expected behavior of the operator? or, how does the operator manage the mapping between feature-gate and components? a feasible way is:

```golang
var (
  // listing global feature gates here
  Foo featuregate = {phase: alpha, default: false}
  Bar featuregate = {phase: beta, default: true}
)

// manage the mapping here
var componentApplicableFeatureGates = map[<component name>][]featuregate
```